### PR TITLE
Initial internal API

### PR DESCRIPTION
**Pull request by jeffsegal-ccpo**
This initial internal API will power the Dashboard & Provisioning Wizard screens.

Does not yet include:
- email invitations
- comprehensive example data for mock services

Replaces #15

Ticket: AT-6363

### CODE REVIEW

File: provisioning/atat_provisioning_api.yaml [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+          description: Portfiolio Draft with the given ID does not exist
+    delete:
+      tags:
+        - portfolioDraft
+      description: Deletes a Portfolio Draft
+      operationId: deletePortfolioDraft
+      parameters:
+        - name: portfolioDraftId
+          in: path
+          description: ID of the Portfolio
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:86]:
```suggestion
          description: ID of the Portfolio Draft
```

File: provisioning/atat_provisioning_api.yaml [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+  - name: portfolioDraft
+    description: >-
+      Portfolio Drafts represent instances of portfolio provisioning workloads which have not yet been submitted to
+      the CSP
+paths:
+  /portfolioDrafts:
+    get:
+      tags:
+        - portfolioDraft
+      description: Gets all Portfolio Drafts the caller has read access to
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:19]:
```suggestion
      description: Gets all Portfolio Drafts to which the caller has read access
```

File: provisioning/atat_provisioning_api.yaml [hunk 3]
```diff
@@ -0,0 +1,524 @@
+openapi: 3.0.2
+info:
+  description: >-
+    This is a DRAFT version of the ATAT Portfolio Draft API
+  version: 1.0.0
+  title: ATAT Internal API - Portfolio Drafts
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:6]:
```suggestion
  title: ATAT Portfolio Draft API
```

**Comment by jeffsegal-ccpo** [comment]
To @zachclark-ccpo 's comment in https://github.com/dod-ccpo/atat-web-api/pull/15 on the empty POST bodies - I intentionally left them empty for the "create new draft" and "submit draft" actions because there was no actual data necessary to transmit there. We could technically create an EmptyBody schema object but that feels like it might violate KISS a bit.

File: provisioning/atat_provisioning_api.yaml [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+        - $ref: '#/components/parameters/limitParam'
+      responses:
+        '200':
+          description: Successful operation with empty response body
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PortfolioDraftSummary'
+        '404':
+          description: Portfolio step not found for this Portfolio Draft or the given Portfolio Draft does not exist
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:32]:
```suggestion
          description: Portfolio Step not found for this Portfolio Draft or the given Portfolio Draft does not exist
```

File: provisioning/atat_provisioning_api.yaml [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+            type: string
+      responses:
+        '200':
+          description: Successful operation with empty response body
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PortfolioStep'
+        '404':
+          description: Portfolio step not found for this Portfolio Draft
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:116]:
```suggestion
          description: Portfolio Step not found for this Portfolio Draft
```

File: provisioning/atat_provisioning_api.yaml [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
+            type: string
+      responses:
+        '200':
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ApplicationStep'
+        '404':
+          description: Application step not found for this Portfolio Draft
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:230]:
```suggestion
          description: Application Step not found for this Portfolio Draft
```

File: provisioning/atat_provisioning_api.yaml [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ApplicationStep'
+  '/portfolioDrafts/{portfolioDraftId}/submit':
+    post:
+      tags:
+        - portfolioDraft
+      description: Submits the given Portfolio Draft for provisioning with a CSP
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:269]:
```suggestion
      description: Submits the given Portfolio Draft for provisioning to a CSP
```

File: provisioning/atat_provisioning_api.yaml [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+      operationId: submitPortfolioDraft
+      parameters:
+        - name: portfolioDraftId
+          in: path
+          description: ID of the Portfolio
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:274]:
```suggestion
          description: ID of the Portfolio Draft
```

File: provisioning/atat_provisioning_api.yaml [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: Successful operation
+        '403':
+          description: User is not authorized to submit this portfolio draft
```
Comment (jasonburkert-ccpo) [inline provisioning/atat_provisioning_api.yaml:282]:
```suggestion
          description: User is not authorized to submit this Portfolio Draft
```

File: provisioning/atat_provisioning_api.yaml (hunk 1 above)
Comment (jeffsegal-ccpo) [reply provisioning/atat_provisioning_api.yaml:86]: fixed

File: provisioning/atat_provisioning_api.yaml (hunk 3 above)
Comment (jeffsegal-ccpo) [reply provisioning/atat_provisioning_api.yaml:6]: @whit1206 had asked for `Internal` to be there so I'll leave it in unless he is OK with your suggestion

**Comment by zachclark-ccpo** [comment]
> To @zachclark-ccpo 's comment in #15 on the empty POST bodies - I intentionally left them empty for the "create new draft" and "submit draft" actions because there was no actual data necessary to transmit there. We could technically create an EmptyBody schema object but that feels like it might violate KISS a bit.

Understandable, thank you.

**Review by zachclark-ccpo — APPROVED** [review]

File: provisioning/atat_provisioning_api.yaml [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+          schema:
+            type: string
+      responses:
+        '200':
+          description: Successful operation with empty response body
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PortfolioStep'
+        '405':
```
Comment (kylelaker-ccpo) [inline provisioning/atat_provisioning_api.yaml:138]: Is 405 the right status code? We allow POSTs but the client failed to make a valid request. Would 400 be more appropriate? Any scenario where a 405 feels appropriate would be handled by the API management service.

File: provisioning/atat_provisioning_api.yaml [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+      responses:
+        '200':
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/FundingStep'
+        '404':
+          description: Funding step not found for this Portfolio Draft
+        '504':
```
Comment (kylelaker-ccpo) [inline provisioning/atat_provisioning_api.yaml:173]: 504 does not feel like an appropriate response for a client error.

File: provisioning/atat_provisioning_api.yaml [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+                $ref: '#/components/schemas/PortfolioDraftSummary'
+        '404':
+          description: Portfolio step not found for this Portfolio Draft or the given Portfolio Draft does not exist
+    post:
+      tags:
+        - portfolioDraft
+      description: Creates a new Portfolio Draft
+      operationId: createPortfolioDraft
+      responses:
+        '200':
```
Comment (kylelaker-ccpo) [inline provisioning/atat_provisioning_api.yaml:39]: A few of these `200`s on a POST should probably be `201`s as well if it results in resource creation.

File: provisioning/atat_provisioning_api.yaml [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+          description: Successful operation with empty response body
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PortfolioStep'
+        '404':
+          description: Portfolio step not found for this Portfolio Draft
+        '504':
```
Comment (anhtran-ccpo) [inline provisioning/atat_provisioning_api.yaml:117]: Is 504 the right code for this?

File: provisioning/atat_provisioning_api.yaml [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+      responses:
+        '200':
+          description: Successful operation
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ApplicationStep'
+        '404':
+          description: Application step not found for this Portfolio Draft
+        '504':
```
Comment (anhtran-ccpo) [inline provisioning/atat_provisioning_api.yaml:231]: Similar to my other comment on the 504

**Review by jasonburkert-ccpo — APPROVED** [review]
Looks good.  See comments about status codes from other reviewers.

File: provisioning/atat_provisioning_api.yaml (hunk 10 above)
Comment (jeffsegal-ccpo) [reply provisioning/atat_provisioning_api.yaml:138]: Yeah that's probably fair. Will update.

File: provisioning/atat_provisioning_api.yaml (hunk 11 above)
Comment (jeffsegal-ccpo) [reply provisioning/atat_provisioning_api.yaml:173]: Ah yes of course. I was trying to be sneaky and add some context for when the first element in the URI path does not exist (vs when the terminal element does, i.e. `/funding`).

But we can probably just make it a 400 when the underlying draft does not exist.

File: provisioning/atat_provisioning_api.yaml (hunk 12 above)
Comment (jeffsegal-ccpo) [reply provisioning/atat_provisioning_api.yaml:39]: Woops, good catch. Should all be fixed now.

**Pull request merged**